lib/web/machineid.go

```diff
@@ -210,3 +211,57 @@ func (h *Handler) getBot(w http.ResponseWriter, r *http.Request, p httprouter.Pa
 
 	return bot, nil
 }
+
+// updateBot updates a bot with provided roles. The only supported change via this endpoint today is roles.
+func (h *Handler) updateBot(w http.ResponseWriter, r *http.Request, p httprouter.Params, sctx *SessionContext, site reversetunnelclient.RemoteSite) (interface{}, error) {
+	var request updateBotRequest
+	if err := httplib.ReadJSON(r, &request); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	botName := p.ByName("name")
+	if botName == "" {
+		return nil, trace.BadParameter("empty name")
+	}
+
+	clt, err := sctx.GetUserClient(r.Context(), site)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	bot, err := clt.BotServiceClient().GetBot(r.Context(), &machineidv1.GetBotRequest{
+		BotName: botName,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err, "unable to find existing bot")
+	}
+
+	mask, err := fieldmaskpb.New(&machineidv1.Bot{}, "spec.roles")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	updated, err := clt.BotServiceClient().UpdateBot(r.Context(), &machineidv1.UpdateBotRequest{
+		UpdateMask: mask,
```

Comment:
Technically, the advantage of this update mask is that you do not need to include the fields below which you are not updating. Only the name, and the spec.roles field need to be provided - the rest /should/ be ignored (and hence we probably shouldn't send them here). This could actually save you from even needing to Get before you Update here.